dashboard/app/models/contact_rollups_pardot_memory.rb

```diff
@@ -31,50 +31,89 @@ def self.add_and_update_pardot_ids(last_id = nil)
     PardotV2.retrieve_new_ids(last_id).each do |mapping|
       pardot_record = find_or_initialize_by(email: mapping[:email])
       pardot_record.pardot_id = mapping[:pardot_id]
-      pardot_record.pardot_id_updated_at = Time.now
+      pardot_record.pardot_id_updated_at = Time.now.utc
       pardot_record.save
     end
   end
 
   def self.create_new_pardot_prospects
+    # Adds contacts to a batch and then sends batch requests to create new Pardot prospects.
+    # Requests may not be sent immediately until batch size is big enough.
+    pardot_writer = PardotV2.new
+    ActiveRecord::Base.connection.exec_query(query_new_contacts).each do |record|
+      data = JSON.parse(record['data']).deep_symbolize_keys
+      submissions, errors = pardot_writer.batch_create_prospects record['email'], data
+      save_sync_results(submissions, errors, Time.now.utc) if submissions.present?
+    end
+
+    # There could be prospects left in the batch because batch size is not yet big enough
+    # to trigger a Pardot request. Sends the remaining of the batch to Pardot now.
+    submissions, errors = pardot_writer.batch_create_remaining_prospects
+    save_sync_results(submissions, errors, Time.now.utc) if submissions.present?
+  end
+
+  def self.update_pardot_prospects
+    pardot_writer = PardotV2.new
+    ActiveRecord::Base.connection.exec_query(query_updated_contacts).each do |record|
+      # If pardot_id has changed since the last data sync, we should assume that
+      # Pardot prospect data is currently empty and re-sync all contact data.
+      old_prospect_data =
+        record['pardot_id_changed'] ?
+          {} :
+          JSON.parse(record['data_synced'] || '{}').deep_symbolize_keys
+      new_contact_data = JSON.parse(record['data']).deep_symbolize_keys
+
+      submissions, errors = pardot_writer.batch_update_prospects(
+        record['email'],
+        record['pardot_id'],
+        old_prospect_data,
```

Comment:
I like referring to Pardot data as "prospect" and Code.org as "contact", I'll try to be consistent about that as well.